### PR TITLE
Use availabelDevice On Electron

### PR DESCRIPTION
**Pull request by lionkunonly**
#17

### CODE REVIEW

**Comment by lionkunonly** [comment]
To use Interface `Core` in UI.js, I use the `Inference-Engine-Node` in `ui.js` instead of 'render.js'. I am not sure if it is good to do it. Do you have any suggestions?

**Comment by huningxin** [comment]
renderer.js and ui.js are all loaded into Electron renderer process. There is no difference where to call `const ie = require('inference-engine-node');`.

File: example/hello_classification_electron/renderer.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-const ie = require('inference-engine-node');
+// const ie = require('inference-engine-node');
```
Comment (huningxin) [inline example/hello_classification_electron/renderer.js]: remove the commented line if it is not needed.

File: example/hello_classification_electron/ui.js [hunk 2]
```diff
@@ -16,26 +18,26 @@ if (!location.search) {
...
-      us = 'camera';
-      strsearch = `?s=${us}`;
-      window.history.pushState(null, null, strsearch);
-      updateScenario(true);
-    });
+  $('#cam').click(() => {
+    us = 'camera';
+    strsearch = `?s=${us}`;
+    window.history.pushState(null, null, strsearch);
+    updateScenario(true);
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: Are above lines only for coding style change? If they are, please group them into a separated commit to fix coding style.

File: example/hello_classification_electron/ui.js [hunk 3]
```diff
@@ -72,10 +74,13 @@ $(document).ready(() => {
 });
 
 $(document).ready(() => {
-  if (hasUrlParam('plugin')) {
-    $('#' + up).attr('checked', 'checked');
-    $('#l-' + up).addClass('checked');
-  }
+  availableDevice = ie.createCore().getAvailableDevices()
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: Please avoid defining variable without `var`, `let` or `const`.

Comment (huningxin) [reply example/hello_classification_electron/ui.js]: `availableDevices`?

File: example/hello_classification_electron/ui.js [hunk 4]
```diff
@@ -72,10 +74,13 @@ $(document).ready(() => {
...
+  console.log('Get availabelDevice')
+  console.log(availableDevice)
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: If above two debugging lines are not used, please remove them.

File: example/hello_classification_electron/ui.js [hunk 5]
```diff
@@ -72,10 +74,13 @@ $(document).ready(() => {
...
+  // The defalut Device is the first in the availableDevice
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: Please move this comment to `constructDevice` method.

File: example/hello_classification_electron/ui.js [hunk 6]
```diff
@@ -116,19 +121,19 @@ $(document).ready(() => {
...
+  ImageElement.addEventListener('load', () => {
     if (us === 'image') {
       predictImage(imageElement);
     }
   });
 });
 
 $(window).ready(() => {
-  currentDevice = up;
+  // currentDevice = up;
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: remove commented code.

File: example/hello_classification_electron/utils.js [hunk 7]
```diff
@@ -13,7 +13,7 @@ const hasUrlParam =
 let ub = getUrlParam('b');
 let us = getUrlParam('s');
 let um = getUrlParam('m');
-let up = getUrlParam('plugin');
+// let up = getUrlParam('plugin');
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: ditto.

File: example/hello_classification_electron/ui.js [hunk 8]
```diff
@@ -116,19 +121,19 @@ $(document).ready(() => {
   InputElement.addEventListener('change', (e) => {
     let files = e.target.files;
     if (files.length > 0 && us === 'image') {
-      imageElement.src = URL.createObjectURL(files[0]);  
+      imageElement.src = URL.createObjectURL(files[0]);
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: move it to coding style commit.

File: example/hello_classification_electron/ui.js [hunk 9]
```diff
@@ -116,19 +121,19 @@ $(document).ready(() => {
...
     }
   }, false);
 
-  ImageElement.addEventListener('load', ()=> {
+  ImageElement.addEventListener('load', () => {
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: ditto.

File: example/hello_classification_electron/utils.js [hunk 10]
```diff
@@ -64,25 +64,24 @@ const disableModel =
...
+          modelList.filter((m) => format === m.format && !m.disabled);
       for (const model of models) {
         const modelName = model.modelName.replace(/ \(.*\)$/, '');
         const modelId = model.modelId;
         tdata.append($(`<input type='radio' class='d-none' name='m' id='${
-              modelId}' value='${modelId}'>`));
+            modelId}' value='${modelId}'>`));
         tdata.append($(`<label id='l-${modelId}' class='themodel' for='${
-              modelId}' title='${modelName}'>${modelName}</label>`));
+            modelId}' title='${modelName}'>${modelName}</label>`));
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: coding style commit.

File: example/hello_classification_electron/utils.js [hunk 11]
```diff
@@ -94,83 +93,108 @@ let singleModelTable =
...
-let getModelClasss = () => {
-  let ids = [];
-  for (let model of $('#query tbody .model')) {
-    ids.push(model.id);
-  }
-  return [...new Set(ids)];
-}
+let constructDevice =
+    (deviceArray) => {
+      console.log('availabelDevice start')
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: please remove this debugging log.

File: example/hello_classification_electron/utils.js [hunk 12]
```diff
@@ -94,83 +93,108 @@ let singleModelTable =
...
-  
-    if (currentBackend && currentDevice) {
-      strsearch = `?b=${currentBackend}&plugin=${currentDevice}&m=${um}&s=${us}`;
-    } else {
-      strsearch = `?b=${ub}&plugin=${up}&m=${um}&s=${us}`;
+
+      // deviceBody.prepend(brows);
+      currentDevice = deviceArray[0].toLowerCase();
+      up = currentDevice
+      console.log(currentDevice)
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: remove this debugging log.

File: example/hello_classification_electron/utils.js [hunk 13]
```diff
@@ -94,83 +93,108 @@ let singleModelTable =
...
-  
-      currentModel = um;
-    }
-  
-    if (currentBackend && currentDevice) {
-      strsearch = `?b=${currentBackend}&plugin=${currentDevice}&m=${um}&s=${us}`;
-    } else {
-      strsearch = `?b=${ub}&plugin=${up}&m=${um}&s=${us}`;
+
+      // deviceBody.prepend(brows);
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: remove the commented code.

File: example/hello_classification_electron/utils.js [hunk 14]
```diff
@@ -94,83 +93,108 @@ let singleModelTable =
...
-  };
-
-const changePlugin = () => {
-    $('.alert').hide();
-    up = $('input:radio[name="bw"]:checked').attr('id');
-
-    if (currentDevice === up || progress) {
-      return;
+
+let getModelClasss =
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js:117]: any changes for this function?

File: example/hello_classification_electron/utils.js [hunk 15]
```diff
@@ -94,83 +93,108 @@ let singleModelTable =
...
-    $('#' + up).attr('checked', 'checked');
-    $('#l-' + up).addClass('checked');
-
-    currentDevice = up;
-    if (currentBackend && currentDevice) {
-      strsearch =
-          `?b=${currentBackend}&plugin=${currentDevice}&m=${um}&s=${us}`;
-    } else {
-      strsearch = `?b=${ub}&plugin=${up}&m=${um}&s=${us}`;
+const changeModel = () => {
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: ditto

**Review by huningxin — COMMENTED** [review]
Thanks for the PR @lionkunonly . This PR mixed both new feature and coding style fix. Please use two commits, one for new feature (use getAvailableDevices) and one for coding style fix. That would be very helpful for review. Thanks!

**Comment by lionkunonly** [comment]
Sorry, I just use the `clang `to fix the coding style? I think some changes are caused by the `clang` and I pay little attention on it. It is my mistake.

**Comment by lionkunonly** [comment]
> Thanks for the PR @lionkunonly . This PR mixed both new feature and coding style fix. Please use two commits, one for new feature (use getAvailableDevices) and one for coding style fix. That would be very helpful for review. Thanks!

Hi, Ningxin. Can I give a PR to change the code style first? Beacause I think it takes some time to change the code style of this PR, and the code style will be changed finally.

**Comment by huningxin** [comment]
> Hi, Ningxin. Can I give a PR to change the code style first?

Yes. Please send another PR for coding style change. We can review and merge that first. Then you can rebase this PR on that.

**Comment by lionkunonly** [comment]
@huningxin  Hi, Ningxin. I think there is no coding style changes in this PR.

File: example/hello_classification_electron/ui.js [hunk 16]
```diff
@@ -72,10 +74,10 @@ $(document).ready(() => {
 });
 
 $(document).ready(() => {
-  if (hasUrlParam('plugin')) {
-    $('#' + up).attr('checked', 'checked');
-    $('#l-' + up).addClass('checked');
-  }
+  let availableDevices = ie.createCore().getAvailableDevices()
```
Comment (huningxin) [inline example/hello_classification_electron/ui.js]: Use `const`?

File: example/hello_classification_electron/utils.js [hunk 17]
```diff
@@ -93,6 +92,26 @@ let singleModelTable =
       return allFormats;
     }
 
+let constructDevice =
```
Comment (huningxin) [inline example/hello_classification_electron/utils.js]: I'd prefer to naming it `configureDevice`. And please put a brief document for this function and note that the first device in the device array will be used.

**Review by huningxin — COMMENTED** [review]
Thanks for the update. I left some review comments, please take a look.

File: example/hello_classification_electron/ui.js (hunk 16 above)
Comment (lionkunonly) [reply example/hello_classification_electron/ui.js]: Done

File: example/hello_classification_electron/utils.js (hunk 17 above)
Comment (lionkunonly) [reply example/hello_classification_electron/utils.js]: Do I need to create a new document at the path `doc\`? Or I just add some introduction in the `READMe.md` in the example path `example\hello_classification_electron\README.md`.

Comment (huningxin) [reply example/hello_classification_electron/utils.js]: doc here means comments besides code.

**Comment by lionkunonly** [comment]
The comment has been added, please review the code.

**Review by huningxin — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**